Better Negative CompoundPeriods

omus opened the pull request:
Currently when making negative compound periods you get odd looking results such as:

```
julia> Dates.Hour(-1) + Dates.Minute(1)
-1 week, 6 days, 23 hours, 1 minute
```

This pull request updates compound periods to be much more human readable:

```
julia> Dates.Hour(-1) + Dates.Minute(1)
-59 minutes
```

omus committed:
Improved support for negative CompoundPeriods.

omus committed:
Fixed associative addition with mixed signs.

nalimilan commented on base/dates/periods.jl:
```diff
@@ -166,10 +166,45 @@ type CompoundPeriod <: AbstractTime
                 end
             end
-            return new(reverse!(pc))
+            # return new(reverse!(pc))
```
Why did you keep this code in a comment?

omus replied:
I should have deleted that line. I'll clean that up.

omus committed:
Removed commented out code.

quinnj commented:
@stevengj mind taking a look?

stevengj commented on base/dates/periods.jl:
```diff
@@ -166,10 +166,44 @@ type CompoundPeriod <: AbstractTime
                 end
             end
-            return new(reverse!(pc))
+            p = reverse(pc)
```
why not `reverse!(pc)` to save the allocation?

omus replied:
I totally agree with you. Not sure what I was thinking

stevengj commented on test/dates/periods.jl:
```diff
@@ -288,6 +288,8 @@ emptyperiod = ((y + d) - d) - y
 @test h + 3mi == 63mi
 @test y - m == 11m
+@test -h + mi == -59mi
+@test string(-h + mi) == "-59 minutes"
```
would be good to have tests for more than two operands

stevengj commented on base/dates/periods.jl:
```diff
@@ -166,10 +166,44 @@ type CompoundPeriod <: AbstractTime
+            i = n - 1
+            while true
+                Pc, f = coarserperiod(P)
+                if i > 0 && typeof(p[i]) == P
```
why should this branch ever be taken?  shouldn't the previous loop have already coalesced equal period types?

stevengj replied:
Oh, nevermind, I see that the loop itself re-introduces duplicates.

stevengj commented on base/dates/periods.jl:
```diff
@@ -166,10 +166,44 @@ type CompoundPeriod <: AbstractTime
             return new(resize!(p, n))
         end
+        # canonicalize Period values such that they are all either
+        # positive or negative.
```
I don't think this is actually possible.  e.g. consider `Year(-1) + Day(1)`.  Since `Year` is not a fixed number of days, it is not possible to make it all positive all negative.

omus replied:
Thats a good point. I'll revise my pull request as your example doesn't behave as it should.

omus replied:
@stevengj code should now be working as expected. I also updated the comment to reflect the new behaviour.

omus committed:
Switched to inline reverse function.

omus committed:
Fixed issue with non-convertable periods.

The coarser period function allows for conversions of Milliseconds
to Seconds, Seconds to Minutes, ... and Days to Weeks. Weeks are
not convertable to Months but Months are convertable to Years. This
gap in the allowed conversions was previously not handled by the
canonicalization which attempts to have all of Period to be the
same sign.

stevengj commented on base/dates/periods.jl:
```diff
@@ -166,10 +166,55 @@ type CompoundPeriod <: AbstractTime
+        if n > 0
+            pc = sizehint!(Period[], n)
+            i = n
+            while i > 0
```
So the canonicalization needs O(n^2) computation, due to the nested `i` and `j` loops?  I guess `n` is never going to be more than 10 or so, but still this is surprising.

omus replied:
I believe it is just O(n). Each iteration of the outer loop covers deals with a non-overlapping range. Each of those ranges is looped over twice by the interior loops.

stevengj replied:
Oh I see... you're imagining a case where we might have two disjoint sets of fungible periods, like years and centuries vs. hours and minutes.

omus replied:
Exactly. The code allows for `coarserperiod` to change how the disjoint sets work and the code just adjusts accordingly.

omus committed:
Minor changes to comments.

quinnj commented:
I think this is good to go (sorry it fell by the way-side). @omus, would you mind taking one more look over since it's been a while? I'm wondering if we want to add a snippet to the docs as well around the behavior just to clarify.

omus commented:
The code still seems reasonable. Since I originally wrote the code I've come to believe that we shouldn't be automatically combining different period types together but I'll save that conversation for a future PR.

I'll add some documentation to the manual about this behaviour. @quinnj I am correct in that the manual currently has no documentation regarding `CompoundPeriod`s?

omus committed:
Add documentation for CompoundPeriod constructor

omus committed:
Minor changes to documentation

omus commented:
I've rebased the code and added documentation for the `CompoundPeriod` constructor. @quinnj does this documentation make the new behaviour clear? Or were you looking for something else?

stevengj commented on doc/stdlib/dates.rst:
```diff
@@ -535,6 +535,31 @@ Periods
+
+   Construct a ``CompoundPeriod`` from a ``Vector`` of ``Period``\ s. The provided ``Period``\ s will be simplified using the following rules:
+
+   * All ``Period``\ s of the same type will be added together
```
It's important to document what `CompoundPeriod` is *for*.  Maybe something like:

> `CompoundPeriod` is useful for expression time periods that are not a fixed multiple of smaller periods.  For example, "a year and a day" is not a fixed number of days, but can be expressed using `CompoundPeriod`.   In fact, a `CompoundPeriod` is automatically generated by addition of incommensurate periods, e.g. `Dates.Year(1) + Dates.Day(1)` produces a `CompoundPeriod` result. 
> When a `CompoundPeriod` is constructed, it is automatically simplified into a canonical form according to the following rules:

stevengj replied:
Also:
> Because of this canonicalization, `CompoundPeriod` is also useful to convert time periods into more human-recognizable forms.  For example, `Dates.CompoundPeriod(Dates.Minute(50000))` yields 4 weeks, 6 days, 17 hours, and 20 minutes.

omus replied:
Thanks @stevengj. I'll make sure to integrate your suggestions.

omus replied:
I ended up splitting the documentation between the `CompoundPeriod` type and the constructor. Let me know what you think.

omus committed:
Add documentation for CompoundPeriod type

tkelman commented on doc/stdlib/dates.rst:
```diff
@@ -19,6 +19,10 @@ Dates and Time Types
 
    ``Period`` types represent discrete, human representations of time.
 
+.. data:: CompoundPeriod
```
`.. data::` forms don't get auto-populated from docstrings, do they?

omus replied:
Unfortunately no, it won't auto-populate. I couldn't find any examples of how to get type docstrings to auto-populate at the time of writing. I'll investigate this more tomorrow but I think another PR will be required to fix this https://github.com/JuliaLang/julia/issues/15623#issuecomment-201332161
